src/main/java/org/apache/maven/plugin/compiler/AbstractCompilerMojo.java

```diff
@@ -1156,6 +1151,16 @@ public void execute() throws MojoExecutionException, CompilationFailureException
 
             incrementalBuildHelper.beforeRebuildExecution(incrementalBuildHelperRequest);
 
+            // Cleanup the generated sources directory
+            if (getGeneratedSourcesDirectory() != null) {
```

Comment:
Why cleaning this directory? I don't understand the reason. Can you please explain?
reading the change as it it looks this will remove the generated-sources added via the phase `generate-sources` or any phase before compile. 
Maybe I read the change in a wrong way?

Reply:
Yes, let me explain, the generatedSourcesDirectory  is actually the directory for the annotation processor's generated source code:
https://github.com/apache/maven-compiler-plugin/blob/e5375fd10f00e4db5471bdc69ebdc4ba20f7bbdc/src/main/java/org/apache/maven/plugin/compiler/CompilerMojo.java#L114-L122

Since the "incremental" feature is actually a rebuild feature, this directory is generated by the javac compiler, it needs to be cleaned before the rebuild of the module or the build will fail with strange errors, this probably fixes many bugs related to the "incremental" compilation and annotation processors.

Reply:
Oh right the method name or even field is kind of wrong but we can't change that